src/util/logging.cpp

```diff
@@ -1,112 +1,161 @@
+            }
         }
     }
 }
 
-// Debug message handler which outputs to stderr and a logfile, prepending the
-// thread name and log level.
-void MessageHandler(QtMsgType type,
-                    const QMessageLogContext&, const QString& input) {
-    // For "]: " and '\n'.
-    size_t baSize = 4;
-    const char* tag = nullptr;
-    bool shouldPrint = true;
-    bool shouldFlush = false;
+} // anonymous namespace
+
+namespace mixxx {
+
+namespace {
+
+// Debug message handler which outputs to stderr and a logfile,
+// prepending the thread name, log category, and log level.
+void handleMessage(
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& input) {
+    const char* levelName = nullptr;
+    WriteFlags writeFlags = WriteFlag::None;
     bool isDebugAssert = false;
     bool isControllerDebug = false;
     switch (type) {
-        case QtDebugMsg:
-            tag = "Debug [";
-            baSize += strlen(tag);
-            isControllerDebug = input.startsWith(QLatin1String(
-                ControllerDebug::kLogMessagePrefix));
-            shouldPrint = Logging::enabled(LogLevel::Debug) ||
-                    isControllerDebug;
-            shouldFlush = Logging::flushing(LogLevel::Debug);
-            break;
-        case QtInfoMsg:
-            tag = "Info [";
-            baSize += strlen(tag);
-            shouldPrint = Logging::enabled(LogLevel::Info);
-            shouldFlush = Logging::flushing(LogLevel::Info);
-            break;
-        case QtWarningMsg:
-            tag = "Warning [";
-            baSize += strlen(tag);
-            shouldPrint = Logging::enabled(LogLevel::Warning);
-            shouldFlush = Logging::flushing(LogLevel::Warning);
-            break;
-        case QtCriticalMsg:
-            tag = "Critical [";
-            baSize += strlen(tag);
-            shouldFlush = true;
-            isDebugAssert = input.startsWith(QLatin1String(kDebugAssertPrefix));
-            break;
-        case QtFatalMsg:
-            tag = "Fatal [";
-            baSize += strlen(tag);
-            shouldFlush = true;
-            break;
-        default:
-            tag = "Unknown [";
-            baSize += strlen(tag);
+    case QtDebugMsg:
+        levelName = "Debug";
+        isControllerDebug =
+                input.startsWith(QLatin1String(
+                        ControllerDebug::kLogMessagePrefix));
+        if (isControllerDebug ||
+                Logging::enabled(LogLevel::Debug)) {
+            writeFlags |= WriteFlag::StdErr;
+            writeFlags |= WriteFlag::File;
+        }
+        if (Logging::shouldFlush(LogLevel::Debug)) {
+            writeFlags |= WriteFlag::Flush;
+        }
+        // TODO: Remove the following line.
+        // Do not write debug log messages into log file if log level
+        // Debug is not enabled starting with release 2.4.0! Until then
+        // write debug messages unconditionally into the log file
```

Comment:
That would be new. A full log file was helpful in the past.

Reply:
I disagree. This is **NOT** helpful now that the level *Info* is avaialable.

I would like to use and have more debug logs during development. But knowing that all of this spams the log file and let it grow to multiple (10-100) megabytes just for a single session prevents me from doing so.

Log level *Info* should be used instead of *Debug* for the detailed logs as everyone else is doing. We already need a special filter rule, because we are doing something uncommon. Debug should strictly be reserved for developers.

Reply:
OK, That makes sense. Unfortunately this step requires to turn many debug messages into info. 
A bunch of rubber-band work.

Reply:
Some loss is expected. We cannot migrate everything all at once.

Switching the file log level with immediately after branching off 2.3 is crucial to ensure that we have enough time to migrate the most important log messages before the next release.